### ISSUE TITLE
Run after gatsby-plugin-postcss to allow usage with TailwindCSS

### ISSUE THREAD

**Issue by ooloth** [anantoghosh/gatsby-plugin-purgecss#100]
Thanks for creating this!

I use TailwindCSS + PurgeCSS with Gatsby via a custom setup in `gatsby-node.js`. It works just fine, but it would be nice to hide the implementation details in a plugin like this one.

However, I notice in the docs that this plugin cannot be used with postcss because:

> `gatsby-plugin-purgecss` is executed before postcss loader

Is there a reason for this order? If possible, it would be useful to run purgecss after postcss to allow TailwindCSS or other setups that require postcss to run first.

**Comment by anantoghosh** [comment]
> plugin cannot be used with postcss

It does not work with postcss syntax based plugins like `postcss-preset-env`. But something like autoprefixer should and does work (if you modify the gatsby webpack config). I don't know if tailwind does css syntax modifications so I can't comment if that is the issue.
Also it may not work with gatsby-plugin-postcss as it did not exist when I was making this plugin and I have not checked it out.
Could you describe exactly what error did you face?

> Is there a reason for this order?

I initially started with running this plugin after postcss. But at the time, how to include postcss plugins in Gatsby 2 (alpha) wasn't clear. I don't remember the details but this plugin would fail parsing files after postcss, not always but on some projects.
But now that the code is more stable, debugging is easier and `gatsby-plugin-postcss` is available, I'll look at it again.
Do you have any Gatsby project repo with tailwind setup (as a postcss plugin) that I could use?

**Comment by ooloth** [comment]
Yeah, I should have explained: [TailwindCSS](https://tailwindcss.com) is a utility-first (aka. functional, atomic) CSS library. It adds some useful custom directives (e.g. `@apply`) that need to be converted to vanilla CSS syntax by PostCSS before the CSS can be used.

TailwindCSS generates a LOT of utility classes (to make it easier to prototype quickly) and [recommends using PurgeCSS](https://tailwindcss.com/docs/controlling-file-size#removing-unused-css-with-purgecss) to remove the extra ones. That's why it would be handy for `gatsby-plugin-purgecss` to handle this use case.

To use TailwindCSS with Gatsby v2, you just use `gatsby-plugin-postcss` and add TailwindCSS to your postcss pipeline (before `autoprefixer`):

### gatsby-config.js

```js
// ...
plugins: [
  `gatsby-plugin-postcss`,
]
// ...
```

### postcss.config.js

```js
// ...
module.exports = {
  plugins: [
    // ...
    require(`tailwindcss`)(`./src/styles/tailwind.js`),
    require(`autoprefixer`)()
  ]
}
// ...
```

Ideally, PurgeCSS will only run in production (so it doesn't slow down HMR in development). Here's the code (provided by @jquense in [this issue](https://github.com/gatsbyjs/gatsby/issues/5778#issuecomment-402481270)) that gets that working in Gatsby v2:

### gatsby-node.js

```js
const PurgeCssPlugin = require(`purgecss-webpack-plugin`)
const path = require(`path`)
const glob = require(`glob`)

const PATHS = {
  src: path.join(__dirname, `src`)
}

const purgeCssConfig = {
  paths: glob.sync(`${PATHS.src}/**/*.js`, { nodir: true }),
  extractors: [
    {
      // Custom extractor to allow special characters (like ":") in class names
      // See: https://tailwindcss.com/docs/controlling-file-size/#removing-unused-css-with-purgecss
      extractor: class {
        static extract(content) {
          return content.match(/[A-Za-z0-9-_:/]+/g) || []
        }
      },
      extensions: [`js`]
    }
  ],
  whitelist: [],
  whitelistPatterns: [/html/, /body/]
}

exports.onCreateWebpackConfig = ({ actions, stage }) => {
  if (stage.includes(`develop`)) return

  // Add PurgeCSS in production
  if (stage.includes(`build`)) {
    actions.setWebpackConfig({
      plugins: [new PurgeCssPlugin(purgeCssConfig)]
    })
  }
}
```

With this setup, I'm basically good to go on my end, but I wanted to share it in case it helps `gatsby-plugin-purgecss` become a one-stop shop for anyone who wants to use PurgeCSS with Gatsby (including those who need to run PostCSS first).

Let me know if anything above could be improved and feel free to use anything that might help this project.
